#!/usr/bin/R


purpose <- 'Detect the Proximal PA site comparing two conditions.'

##------------------------------------------------------
## setup Rscript
suppressPackageStartupMessages({library(optparse)})

Usage <- paste('%prog [OPTIONS] \n', purpose)
option.list <-
 list(make_option(c('-o', '--outdir'), default = '.', type = 'character',
                 help = 'The directory to output the results to'),
  make_option(c('-c', '--chr'), default = 'all', type = 'character',
                help = 'chr'),
  make_option(c('-p', '--pen.value'), default = 0, type = 'numeric',
                help = 'penalty value for change point'),
  make_option(c('-s', '--sampleinfo'), default = NULL, type = 'character',
                help = 'sample information file with columns named Bam,Seq_depth,Group'),
  make_option(c('-ref', '--reference'), default = NULL, type = 'character',
             help = 'annotated region file generated by Extract_ref_region.R'),
  make_option(c('-n', '--threads'), default = 2, type = 'numeric',
             help = 'number of threads'),
  make_option(c('-nPAs', '--numberPAs'), default = 2, type = 'numeric',
              help = 'the most number of PA sites')
  
)

opt_parser = OptionParser(usage=Usage, option_list=option.list)
opt = parse_args(opt_parser)
if (is.null(opt$ref)){
print_help(opt_parser)
stop("reference file must be supplied .n", call.=FALSE)
}


##--------------------------------------------------------
## script start from here
suppressPackageStartupMessages({
  library(rtracklayer)
  library(data.table)
  library(GenomicRanges)
  library(GenomicFeatures)
  library(annotatr)
  library(dplyr)
  library(Rsamtools)
  library(changepoint)
  library(parallel)
})




##-------------------------------------------------------
#test chrmosome 20
chrs <- opt$chr
if(chrs == 'all'){
  chrs <- c(seq(1,22),'X','Y')
}

########TEST FROM HERE#######################################
bedfile <- opt$reference
anno_file <- fread(bedfile)
sample.info <- opt$sampleinfo
sample.info <- fread(opt$sampleinfo, header=T)
cutoff <- 30
#anno_file <- fread('./ref/GRCh38.3utr.unextended.bed')
#sample.info <- fread('../ALL_sample.info.txt', header=T)
colnames(anno_file) <- c('chr', 'start', 'end','id','score','strand')

UTR_length <- abs(anno_file$start - anno_file$end)

#filter out UTR_length greater than 1e5 ,
#filter out 676 genes,513 genes with UTR length less than200

anno_file <- anno_file[UTR_length<=1e5 &UTR_length >= 200& chr %in%chrs,]

#make Grangelist from data frame
GrList <- makeGRangesListFromDataFrame(anno_file,
                                       keep.extra.columns = TRUE,
                                       names.field = 'id')

flag <- scanBamFlag(isSecondaryAlignment = FALSE,
                    isNotPassingQualityControls = FALSE,
                    isUnmappedQuery = FALSE,
                    #isSecondMateRead = TRUE,
                    isDuplicate = FALSE)

p_param <- PileupParam(distinguish_strands=FALSE,
                       distinguish_nucleotides=FALSE,
                       min_mapq = 1L,
                       max_depth=2e4,
                       min_nucleotide_depth = 0L)



GenerateReadsPerBase <- function(GR, sample.info,flag, p_param, ...) {
  sbp <- ScanBamParam(which=GR, flag=flag)
  #tab <- as.data.table(pileup(BamFile, scanBamParam=sbp, pileupParam=p_param))
  bamFile <- sample.info$Bam
  
  tab <- mclapply(FUN = function(x) as.data.frame(pileup(x, scanBamParam=sbp, pileupParam=p_param)), 
                  bamFile,  
                  mc.cores=5)
  #check if no reads in all sample
  checkreads <- do.call(c,lapply(tab, function(x)nrow(x)))
  if(sum(checkreads) == 0){
    tab.counts <- NULL
  }else{
    tab.counts <- as.data.table(do.call(cbind,(lapply(tab, function(x){
      rownames(x) <- as.character(x$pos)
      x <- x[as.character(seq(start(GR), end(GR),1)),]
      x$count[is.na(x$count)] <- 0
      return(x$count)
    }))))
    #normalize
    tab.counts.n <- t(tab.counts) * (1e6 /sample.info$Seq_depth)
    
    colnames(tab.counts) <- sample.info$sample
    #tab <- tab[strand != as.character(strand(GR)), ]
    tab.counts$Total.reads <- rowSums(tab.counts)
    tab.counts$Normalized.reads <- colSums(tab.counts.n)
    #colnames(tab) <- 'reads'
    
    tab.counts[, pos := seq(start(GR), end(GR), 1)]
    #tab[, seqnames := names(GR)]
    tab.counts[, chr := seqlevelsInUse(GR)]
    
    setkey(tab.counts, pos)
    
  }
  return(tab.counts)
}




##GAM

DetectPAs <- function(GR, cut.off=cutoff){
  cv <- GenerateReadsPerBase(GR=GR,
                             sample.info = sample.info,
                             bamFile.dir = bamFile.dir,
                             #seq_depth = seq_depth,
                             flag = flag,
                             p_param = p_param)
  print(names(GR))
  print(paste0(round(which(anno_file$id == names(GR))/nrow(anno_file),2) * 100,'%',' ','Done'))
  
  if(is.null(cv)|mean(cv$Total.reads) <= cut.off){
    GR.res <- as.data.frame(matrix(NA,1,length(samples)*3 +1))
    colnames(GR.res) <-c(paste0(samples,'.L_Coverage'),paste0(samples,'.S_Coverage'),
                         paste0(samples,'.RUD'),'PA')
    
  }else{
    GR.res <- ChangePoint.RUD(GR=GR,cv=cv)
    
  }
  GR.res$ID <- names(GR)
  return(GR.res)
}

groups <- unique(sample.info$Group)
samples <- sample.info$sample
#Differential percentage of distal APA usage index. We used the following three
#criteria to detect the most significant APA events: first, given long 30-UTR expression level wi
#wi L and short 30-UTR expression level
#S estimated from (equation (1)), we used Fisher’s exact test to determine the P value
#of PDUI difference between tumour and matched normal tissue of the same
#patient, which is further adjusted by the Benjamini–Hochberg procedure to control 
#the false-discovery rate (FDR) at 5%. Second, the absolute mean difference of PDUIs of all the patients in the same tumour type must be no less than 0.2. Third, the mean fold-change of PDUIs of all the patients in the same tumour type must be no less than 1.5


ChangePoint.RUD <- function(GR,cv){
  Start <- start(GR)
  End <- end(GR)
  n <- End - Start+1
  Str <- as.character(strand(GR))
  #at most two PA sites,the cpt.meanvar will out put 3 sites,
  #the last one alway at the last element whic need to be removed.
  PAs <- suppressWarnings(cpt.meanvar(cv$Normalized.reads,method="BinSeg",
                                      test.stat="Exponential",Q=opt$numberPAs,
                                      pen.value = 0.5,penalty='Manual'))
  #Group_A_Mean_PDUI       Group_B_Mean_PDUI       PDUI_Group_diff P_val   adjusted.P_val
  nPAs <- length(PAs@cpts)-1
  if(nPAs >=1){
    PA <- PAs@cpts[1:nPAs]
    PA.df <- as.data.frame(do.call(rbind,lapply(PA,function(PA){
      if(Str == '-'){
        L.cv <- colMeans(cv[1:PA,..samples])
        S.cv <- colMeans(cv[PA:n,..samples])
      }else{
        S.cv <- colMeans(cv[1:PA,..samples])
        L.cv <- colMeans(cv[PA:n,..samples])
      }
      RUD <- L.cv/S.cv
      res <- c(L.cv,S.cv,RUD)
      return(res)
    })))
  }else{
    PA <- NA
    PA.df <- as.data.frame(matrix(NA,1,length(samples)*3))
  }
  colnames(PA.df) <- c(paste0(samples,'.L_Coverage'),paste0(samples,'.S_Coverage'),
                       paste0(samples,'.RUD'))
  PA.df$PA <- Start +PA
  return(PA.df)
}



##all genes on chrs
print('start analysing')


all.res <- do.call(rbind, mclapply(GrList,DetectPAs,mc.cores = opt$threads))

setwd(opt$outdir)
write.table(all.res, file ='IdentifiedPAs.all.txt',
            row.names = F, col.names = T, sep = '\t', quote=F)
            







