#!/usr/bin/R


purpose <- 'Detect the Proximal PA site comparing two conditions.'

##------------------------------------------------------
## setup Rscript
suppressPackageStartupMessages({library(optparse)})

Usage <- paste('%prog [OPTIONS] \n', purpose)
option.list <-
  list(make_option(c('-o', '--outdir'), default = '.', type = 'character',
                   help = 'The directory to output the results to'),
       make_option(c('-b1', '--bam1'), default = NULL, type = 'character',
                   help = 'bamfiles'),
       make_option(c('-b2', '--bam2'), default = NULL, type = 'character',
                   help = 'bamfiles'),
       make_option(c('-c', '--chr'), default = NULL, type = 'character',
                   help = 'chr'),
       make_option(c('-s', '--samplenumber'), default = NULL, type = 'character',
                   help = 'sample number'),
       make_option(c('-B', '--bed'), default = NULL, type = 'character',
                   help = 'annotated region file generated by Ref.annotation.R'),
       make_option(c('-n', '--threads'), default = 5, type = 'numeric',
                   help = 'number of threads')
  )

opt_parser = OptionParser(usage=Usage, option_list=option.list)
opt = parse_args(opt_parser)
if (is.null(opt$bed)){
  print_help(opt_parser)
  stop("bed file must be supplied .n", call.=FALSE)
}


##--------------------------------------------------------
## script start from here
suppressPackageStartupMessages({
  library(rtracklayer)
  library(data.table)
  library(GenomicRanges)
  library(GenomicFeatures)
  library(annotatr)
  library(dplyr)
  library(Rsamtools)
  library(gam)
  library(parallel)
})




##-------------------------------------------------------



#test sample1 groups
sample.number <-  opt$samplenumber
#test chrmosome 20
chrs <- opt$chr

########TEST FROM HERE#######################################
bedfile <- opt$bed
anno_file <- fread(bedfile)
#anno_file <- fread('GRCh38.3utr.unextended.bed')

colnames(anno_file) <- c('chr', 'start', 'end','id','score','strand')

UTR_length <- abs(anno_file$start - anno_file$end)

#filter out UTR_length greater than 1e5 ,
#filter out 676 genes,513 genes with UTR length less than200

anno_file <- anno_file[UTR_length<=1e5 &UTR_length >= 200& chr ==chrs,]

#make Grangelist from data frame
GrList <- makeGRangesListFromDataFrame(anno_file,
                                       keep.extra.columns = TRUE,
                                       names.field = 'id')

flag <- scanBamFlag(isSecondaryAlignment = FALSE,
                    isNotPassingQualityControls = FALSE,
                    isUnmappedQuery = FALSE,
                    #isSecondMateRead = TRUE,
                    isDuplicate = FALSE)


sample.info <- read.table('/mnt/data1/liyupeng/data/SecondHospital/ALL_CTRL/ALL_CTRL_info.csv',
                          header = T, sep = ',', stringsAsFactors = F)
sample.info$samplegroup <- c(seq(1,10),seq(1,10))


sample.info <- sample.info[sample.info$samplegroup == sample.number,]


bamFile.dir <- '/mnt/data1/liyupeng/data/SecondHospital/'
#seq_depth <- fread('Seq_depth',header = T)
#identical(sample.info$ID, seq_depth$ID)

#make sure the order of seq_depth is the same as sample.info####-----

p_param <- PileupParam(distinguish_strands=FALSE,
                       distinguish_nucleotides=FALSE,
                       min_mapq = 1L,
                       include_insertions=T,
                       max_depth=1e4,
                       min_nucleotide_depth = 0L)



GenerateReadsPerBase <- function(GR, sample.info,bamFile.dir, seq_depth,flag, p_param, ...) {
  sbp <- ScanBamParam(which=GR, flag=flag)
  #tab <- as.data.table(pileup(BamFile, scanBamParam=sbp, pileupParam=p_param))
  filedir <- paste0(sample.info$from,
                    '/bam/',
                    sample.info$ID,
                    '.Aligned.sortedByCoord.out.bam')
  
  bamFile <- paste0(bamFile.dir,filedir)
  
  tab <- mclapply(FUN = function(x) as.data.frame(pileup(x, scanBamParam=sbp, pileupParam=p_param)), 
                  bamFile,  
                  mc.cores=2)
  #check if no reads in all sample
  checkreads <- do.call(c,lapply(tab, function(x)nrow(x)))
  if(sum(checkreads) == 0){
    tab.counts <- NULL
  }else{
    tab.counts <- as.data.table(do.call(cbind,(lapply(tab, function(x){
      rownames(x) <- as.character(x$pos)
      x <- x[as.character(seq(start(GR), end(GR),1)),]
      x$count[is.na(x$count)] <- 0
      return(x$count)
    }))))
    #normalize
    #tab.counts.n <- t(t(tab.counts) * 1e6/seq_depth$seq_depth)
    
    colnames(tab.counts) <- sample.info$sample
    #tab <- tab[strand != as.character(strand(GR)), ]
    tab.counts$Total.reads <- rowSums(tab.counts)
    #colnames(tab) <- 'reads'
    
    tab.counts[, pos := seq(start(GR), end(GR), 1)]
    #tab[, seqnames := names(GR)]
    tab.counts[, chr := seqlevelsInUse(GR)]
    setkey(tab.counts, pos)
    
  }
  return(tab.counts)
}


te <- galp0[[1]]
rownames(te) <- te$start
gap_cov <- coverage(galp0)
cut.off <- 30



system.time(cv <- GenerateReadsPerBase(GR=GR,
                           sample.info = sample.info,
                           bamFile.dir = bamFile.dir,
                           #seq_depth = seq_depth,
                           flag = flag,
                           p_param = p_param))

#user  system elapsed 
#9.796   0.370   9.711 

library(GenomicAlignments)
system.time(galp0 <- mclapply(FUN = function(x) as.data.frame(
  readGAlignments(x, use.names = TRUE, param = sbp)),bamFile,mc.cores=2))

system.time(tab <- mclapply(FUN = function(x) as.data.frame(
  pileup(x, scanBamParam=sbp, pileupParam=p_param)), 
                bamFile,  
                mc.cores=2))
#TODO: would apply be faster-----------------------


##GAM

DetectPolyAsite <- function(GR, cut.off){
  cv <- GenerateReadsPerBase(GR=GR,
                             sample.info = sample.info,
                             bamFile.dir = bamFile.dir,
                             #seq_depth = seq_depth,
                             flag = flag,
                             p_param = p_param)
  if(is.null(cv)){
    GR.res <- data.frame( PA=NA, L.r.group1=NA, R.r.group1=NA,
                          L.r.group2=NA, R.r.group2=NA,
                          psi.group1=NA, psi.group2=NA,
                          log2FC=NA,  p_value=NA)
    
  }else{
    cv <- as.data.frame(cv)
    GR.res <- LogLik_gam(cv,GR,30,sample.number = sample)
    
  }
  GR.res$ID <- names(GR)
  return(GR.res)
}




LogLik_gam <- function(cv,GR,cut.off,sample.number){
  s <- sample.number
  cv <- as.data.frame(cv)
  regionReadsBase <- cv[,1]+cv[,2]
  
  POS <- names(GR)
  Start <- start(GR)
  End <- end(GR)
  Strand <- as.character(strand(GR))
  #regression 
  n <- End - Start
  point <- seq(100, n, 20 )

  if(sum(regionReadsBase) > cut.off){
  Loglikeli.points <- mclapply(FUN =function(x){
      mean_abundence = c(rep(mean(regionReadsBase[1:x]), x),
                         rep(mean(regionReadsBase[(x+1):n]), n - x+1))
      fit <- gam(regionReadsBase~mean_abundence)
      loglikeli <- logLik(fit)
      return(loglikeli)
    },point,mc.cores = 5)
  
    po <- point[which.max(Loglikeli.points)]
    selected <- Start + po
    R.r.group1 <- mean(as.numeric(cv[po:n,1]))
    L.r.group1 <- mean(as.numeric(cv[1:po,1]))
    
    R.r.group2 <- mean(as.numeric(cv[po:n,2]))
    L.r.group2 <- mean(as.numeric(cv[1:po,2]))
    
    df <- data.frame(PA = selected, 
                     L.r.group1 = L.r.group1, R.r.group1 = R.r.group1,
                     L.r.group2 = L.r.group2, R.r.group2 = R.r.group2,
                     psi.group1 = ifelse(Strand == '-', 
                                         R.r.group1/L.r.group1, L.r.group1/R.r.group1),
                     psi.group2 = ifelse(Strand == '-',
                                         R.r.group2/L.r.group2, L.r.group2/R.r.group2))
    #log2FC
    df$log2FC <- log2(df$psi.group1/df$psi.group2)
    df$p_value<-
      fisher.test(matrix(c(L.r.group1,L.r.group2,R.r.group1,R.r.group2),
                         nrow = 2))$p.value
  }
  else{
    df <- data.frame( PA=NA, L.r.group1=NA, R.r.group1=NA,
                      L.r.group2=NA, R.r.group2=NA,
                      psi.group1=NA, psi.group2=NA,
                      log2FC=NA,  p_value=NA)
  }
  return(df)
}

##all genes on chrs
print('start analysing')
#ENSG00000184678|HIST2H2BE|1|-
#GR<- GrList[[1478]]
#pdf('Loglikelihood.pdf',6,6)
#plot(regionReadsBase,type='l', xlab = 'Base position of UTR',ylab='Coverage')
#plot(Loglikeli.points,type='l',xlab ='Tested points',ylab='Log Likelihood of GAM')
#dev.off()

all.res <- do.call(rbind, mclapply(GrList,DetectPolyAsite,mc.cores = opt$threads))
setwd(opt$outdir)
write.table(all.res, file = paste0('IdentifiedAPA_sample',opt$samplenumber,'_chr',opt$chr,'.txt'),
            row.names = F, col.names = T, sep = '\t', quote=F)



LogLik_gam <- function(cv,GR,cut.off,sample.number){
  s <- sample.number
  cv <- as.data.frame(cv)
  regionReadsBase <- cv[,1]+cv[,2]
  
  POS <- names(GR)
  Start <- start(GR)
  End <- end(GR)
  Strand <- as.character(strand(GR))
  #regression 
  n <- End - Start
  point <- seq(100, n, 20 )
  Loglikeli.points <- c()
  system.time(galp0 <- mclapply(FUN = function(x) as.data.frame(
    readGAlignments(x, use.names = TRUE, param = sbp)),bamFile,mc.cores=2))
  
  system.time(te <- mclapply(FUN =function(x){
    mean_abundence = c(rep(mean(regionReadsBase[1:x]), x),
              rep(mean(regionReadsBase[(x+1):n]), n - x+1))
    fit <- gam(regionReadsBase~mean_abundence)
    loglikeli <- logLik(fit)
    return(loglikeli)
  },point,mc.cores = 5))
  
  
  if(sum(regionReadsBase) > cut.off){
    system.time(for( i in 1:length(point)){
      df <- data.frame(y = regionReadsBase, 
                       x = c(rep(mean(regionReadsBase[1:point[i]]), point[i]),
                             rep(mean(regionReadsBase[(point[i]+1):n]), n - point[i]+1)))
      fit <- gam(y ~ x, data = df)
      # p[i] <- summary(fit)[4][[1]][1,5] #p_value
      Loglikeli.points[i] <- logLik(fit)
      #p[i] <- AIC(fit)
    })
    po <- point[which.max(Loglikeli.points)]
    selected <- Start + po
    L.r.group1 <- mean(as.numeric(cv[po:n,1]))
    R.r.group1 <- mean(as.numeric(cv[1:po,1]))
    
    L.r.group2 <- mean(as.numeric(cv[po:n,2]))
    R.r.group2 <- mean(as.numeric(cv[1:po,2]))
    
    df <- data.frame(PA = selected, 
                     L.r.group1 = L.r.group1, R.r.group1 = R.r.group1,
                     L.r.group2 = L.r.group2, R.r.group2 = R.r.group2,
                     psi.group1 = ifelse(Strand == '-', 
                                         R.r.group1/L.r.group1, L.r.group1/R.r.group1),
                     psi.group2 = ifelse(Strand == '-',
                                         R.r.group2/L.r.group2, L.r.group2/R.r.group2))
    #log2FC
    df$log2FC <- log2(df$psi.group1/df$psi.group2)
    df$p_value<-
      fisher.test(matrix(c(L.r.group1,L.r.group2,R.r.group1,R.r.group2),
                         nrow = 2))$p.value
  }
  else{
    df <- data.frame( PA=NA, L.r.group1=NA, R.r.group1=NA,
                      L.r.group2=NA, R.r.group2=NA,
                      psi.group1=NA, psi.group2=NA,
                      log2FC=NA,  p_value=NA)
  }
  return(df)
}



