#!/usr/bin/R


purpose <- 'Detect the Proximal PA sites.'

##------------------------------------------------------
## setup Rscript
suppressPackageStartupMessages({library(optparse)})

Usage <- paste('%prog [OPTIONS] \n', purpose)
option.list <-
  list(make_option(c('-o', '--outdir'), default = '.', type = 'character',
                   help = 'The directory to output the results to'),
       make_option(c('-c', '--chr'), default = 'all', type = 'character',
                   help = 'chr'),
       make_option(c('-p', '--pen.value'), default = 0.5, type = 'numeric',
                   help = 'penalty value for change point'),
       make_option(c('-s', '--sampleinfo'), default = NULL, type = 'character',
                   help = 'sample information file with columns named Bam,Seq_depth,Group,sample'),
       make_option(c('-r', '--reference'), default = NULL, type = 'character',
                   help = 'annotated region file generated by Extract_ref_region.R'),
       make_option(c('-n', '--threads'), default = 2, type = 'numeric',
                   help = 'number of threads'),
       make_option(c('-N', '--numberPAs'), default = 2, type = 'numeric',
                   help = 'the most number of PA sites'),
       make_option(c('-l','--utrlength',default=NULL,type='character',
                     help =' length of 3utr for ensembl gene'))
       
  )

opt_parser = OptionParser(usage=Usage, option_list=option.list)
opt = parse_args(opt_parser)
if (is.null(opt$ref)){
  print_help(opt_parser)
  stop("reference file must be supplied .n", call.=FALSE)
}



##--------------------------------------------------------
## script start from here
suppressPackageStartupMessages({
  library(rtracklayer)
  library(data.table)
  library(GenomicRanges)
  library(GenomicFeatures)
  library(annotatr)
  library(dplyr)
  library(changepoint)
  library(parallel)
  library(Rsamtools)
})




##-------------------------------------------------------
#------------
#setwd('/mnt/raid61/Personal_data/xiaoxia/APA/Project/Hema_mouse/bulk_tianjin/src/')
chrs <- '19'
opt$reference <- 'reduced.3utr.txt'
opt$sampleinfo <- '/mnt/raid61/Personal_data/xiaoxia/APA/Project/ALL_apa/ALL_sample.info.txt'
opt$utrlength <- 'utr_length.txt'
#------------


chrs <- opt$chr
if(chrs == 'all'){
  chrs <- c(seq(1,19),'X','Y')
}

anno_file <- fread(opt$reference)
anno_file <- anno_file[seqnames%in%chrs,]
sample.info <- fread(opt$sampleinfo, header=T)
utr_length <- fread(opt$utrlength)
cutoff <- 30
#filter out length of 3utr less than 200nt or >2e4
genes.use <- utr_length[length_3utr >200 & length_3utr <2e4,]$gene_id
anno_file <- anno_file[gene_id%in%genes.use,]
anno_file$use.names <- anno_file$gene_id
#make Grangelist from data frame
GrList <- makeGRangesListFromDataFrame(anno_file,
                                       keep.extra.columns = TRUE,
                                       split.field = 'gene_id')

flag <- scanBamFlag(isSecondaryAlignment = FALSE,
                    isNotPassingQualityControls = FALSE,
                    isUnmappedQuery = FALSE,
                    isPaired = TRUE,
                    isProperPair = TRUE,
                    isDuplicate = FALSE)
p_param <- PileupParam(distinguish_strands=FALSE,
                       distinguish_nucleotides=FALSE,
                       min_mapq = 1L,
                       max_depth=2e4,
                       min_nucleotide_depth = 0L)

bamFile <- sample.info$Bam
samples <- sample.info$sample

GenerateReadsPerBase <- function(GR, sample.info,flag, p_param, ...) {
  sbp <- ScanBamParam(which=GR, flag=flag,mapqFilter = 255)
  bamFile <- sample.info$Bam
  tab <- mclapply(FUN = function(x) as.data.frame(pileup(x, scanBamParam=sbp, pileupParam=p_param)), 
                  bamFile,  
                  mc.cores=5)
  #check if no reads in all sample
  checkreads <- do.call(c,lapply(tab, function(x)nrow(x)))
  if(sum(checkreads) == 0){
    tab.counts <- NULL
  }else{
    tab.tb <- lapply(tab, function(x)x[,2:3])
    tab.counts <- Reduce(function(x,y)merge(x,y,by='pos',all=T),tab.tb)
    tab.counts[is.na(tab.counts)] <- 0
    colnames(tab.counts)[-1] <- sample.info$sample
    tab.counts.n <- t(tab.counts[,-1]) * (1e6 /sample.info$Seq_depth)
    tab.counts$Total.reads <- rowSums(tab.counts[,-1])
    tab.counts$Normalized.reads <- colSums(tab.counts.n)
  }
  return(tab.counts)
}

#changepoint
DetectPAs <- function(GR, cut.off=cutoff){
  #print(unique(GR$use.names))
  cv <- GenerateReadsPerBase(GR=GR,
                             sample.info = sample.info,
                             flag = flag,
                             p_param = p_param)
  if(is.null(cv)|mean(cv$Total.reads) <= cut.off){
    GR.res <- as.data.frame(matrix(NA,1,length(samples)*3 +1))
    colnames(GR.res) <-c(paste0(samples,'.L_Coverage'),paste0(samples,'.S_Coverage'),
                         paste0(samples,'.RUD'),'PA')
    GR.res$gene_id <- unique(GR$use.names)
  }else{
    GR.res <- ChangePoint.RUD(GR=GR,cv=cv)
  }
  GR.res$ID <- names(GR)
  return(GR.res)
}

#GR <- GrList$ENSG00000141968 ALL_Ctrl
pdf('/mnt/raid61/Personal_data/xiaoxia/APA/Project/Hema_hs/src/example.cp.pdf',6,6)
plot(cv$Normalized.reads,type='l',xaxt = 'n',xlab='',ylab='Normalized reads of all samples')
axis(1, at=c(1,100,183,259), labels=c(cv$pos[1],cv$pos[100],cv$pos[183],cv$pos[259]))
abline(v=PAs@cpts)
dev.off()
#
ChangePoint.RUD <- function(GR,cv){
  Str <- unique(as.character(strand(GR)))
  n <- nrow(cv)
  #at most two PA sites,the cpt.meanvar will out put 3 sites,
  #the last one alway at the last element which need to be removed.
  PAs <- suppressWarnings(cpt.meanvar(cv$Normalized.reads,method="BinSeg",
                                      test.stat="Exponential",Q=2,
                                      pen.value = 0.5,penalty='Manual'))
  #Group_A_Mean_PDUI       Group_B_Mean_PDUI       PDUI_Group_diff P_val   adjusted.P_val
  nPAs <- length(PAs@cpts)-1
  if(nPAs >=1){
    PA <- PAs@cpts[1:nPAs]
    PA.df <- as.data.frame(do.call(rbind,lapply(PA,function(PA){
      if(Str == '-'){
        L.cv <- colMeans(cv[1:PA,samples])
        S.cv <- colMeans(cv[PA:n,samples])
      }else{
        S.cv <- colMeans(cv[1:PA,samples])
        L.cv <- colMeans(cv[PA:n,samples])
      }
      RUD <- L.cv/S.cv
      res <- c(L.cv,S.cv,RUD)
      return(res)
    })))
  }else{
    PA <- 1
    PA.df <- as.data.frame(matrix(NA,1,length(samples)*3))
  }
  colnames(PA.df) <- c(paste0(samples,'.L_Coverage'),paste0(samples,'.S_Coverage'),
                       paste0(samples,'.RUD'))
  PA.df$PA <- cv$pos[PA]
  PA.df$gene_id <- unique(GR$use.names)
  return(PA.df)
}




print(chrs)
#all.res <- do.call(rbind, lapply(GrList,DetectPAs))
pb <- txtProgressBar(min = 0, max = length(GrList), style = 3)
all.res <- do.call(rbind,lapply(1:length(GrList), function(i){
  Sys.sleep(0.1)
  setTxtProgressBar(pb, i)
  res <- DetectPAs(GrList[[i]])
  return(res)
}))

close(pb)

setwd(opt$outdir)
write.table(all.res, file =paste0('IdentifiedPAs.',chrs,'.txt'),
            row.names = T, col.names = T, sep = '\t', quote=F)














